File: android/app/src/main/java/buzz/armada/app/db/Dm17.kt
package buzz.armada.app.db

/**
 * The NIP-17 opened-DM store's rules, in Kotlin — a port of the parts of
 * `src/lib/nip17/dm17Store.ts` and `protocol.ts` that decide what may be stored
 * and how it is shaped.
 *
 * The background service decrypts DM gift wraps for its notifications anyway, so
 * it files the recovered rumor in the same per-viewer tenant the WebView reads
 * (`dm17:<self>`) rather than throwing it away and leaving the app to open the
 * wrap a second time. That only works if it obeys the same rules — otherwise the
 * two writers disagree about what a conversation contains.
 *
 * Pure and Context-free so the rules can be tested on the JVM without an
 * emulator; [ServiceStore.storeDm17Rumor] is the thin part that writes.
 */
internal object Dm17 {

    /**
     * The rumor kinds the DM store holds: NIP-09 delete, NIP-25 reaction,
     * chat, file, and Armada's disappearing-messages timer.
     *
     * Typing indicators (23311) are deliberately absent — the signal IS the
     * event's existence, it lives for seconds, and it must never reach the store.
     */
    val KINDS = setOf(5, 7, 14, 15, 1740)

    /**
     * The kinds worth LISTING — a chat message or a file — which [MSG_TERM] and
     * [MINE_TERM] cover. The port of `DM_MESSAGE_KINDS`.
     */
    val MESSAGE_KINDS = setOf(14, 15)

    /** The ArmadaDB tenant holding one viewer's opened DMs. */
    fun tenant(self: String): String = TENANT_PREFIX + self

    /**
     * The NIP-40 deadline these tags carry, or null. `Number(raw)` semantics:
     * anything that isn't a finite number is no deadline at all.
     */
    fun expirationOf(tags: List<List<String?>>): Long? {
        val raw = tags.firstOrNull { it.getOrNull(0) == "expiration" }?.getOrNull(1) ?: return null
        val secs = raw.trim().toDoubleOrNull() ?: return null
        if (secs.isNaN() || secs.isInfinite()) return null
        return secs.toLong()
    }

    /** Whether these tags carry a NIP-40 deadline that has already passed. */
    fun isExpired(tags: List<List<String?>>, now: Long): Boolean {
        val at = expirationOf(tags) ?: return false
        return at <= now
    }

    /**
     * The participants of a rumor's conversation, from `self`'s perspective:
     * everyone involved except the viewer, sorted. `[self]` for Note to Self.
     * Null when unattributable — an own copy with no `p` tag names no room, and
     * callers drop it rather than guess.
     *
     * A port of `dmPeersOf` in `src/lib/nip17/conversation.ts`, and it has to
     * agree with it exactly: this is what [convTerm] files a rumor under, and a
     * rumor filed under a term the WebView does not look up is a message that
     * arrived while the app was dead and is then invisible in the thread.
     *
     * A NIP-17 conversation is its participant SET, not a peer — the `p` set
     * defines the room. Canonicalizing to "everyone but the viewer, sorted"
     * makes the two directions agree: a message Alice sends to {me, Bob} arrives
     * as `pubkey: Alice, p: [me, Bob]` and my reply leaves as
     * `pubkey: me, p: [Alice, Bob]`, and both reduce to [Alice, Bob]. For a 1:1
     * it yields exactly `[peer]`, which is why the keys did not change when
     * groups arrived.
     */
    fun peersOf(rumor: Rumor, self: String): List<String>? {
        val recipients = LinkedHashSet<String>()
        for (tag in rumor.tags) {
            val value = tag.getOrNull(1)
            if (tag.getOrNull(0) == "p" && !value.isNullOrEmpty()) recipients.add(value)
        }

        if (rumor.pubkey != self) {
            // Received: the sender is a participant whether or not they p-tagged
            // themselves, and we are not one of our own peers.
            val others = LinkedHashSet(recipients)
            others.add(rumor.pubkey)
            others.remove(self)
            return if (others.isEmpty()) null else others.sorted()
        }

        // Our own copy: only the `p` set says where it went.
        if (recipients.isEmpty()) return null
        val others = LinkedHashSet(recipients)
        others.remove(self)
        return if (others.isEmpty()) listOf(self) else others.sorted()
    }

    /**
     * Separator between participants in a conversation KEY — the port of
     * `DM_PEER_SEP`. Legal unescaped in a URL path segment, so the deep link a
     * notification carries stays `/dm/<a>,<b>`.
     */
    const val PEER_SEP = ","

    /**
     * The stable string key for a participant set — the port of `dmConvKey`.
     *
     * For a 1:1 (and Note to Self) this is exactly the other party's pubkey, so
     * a `dm:<convKey>` room key, its read marker and its deep link are all
     * byte-identical to the single-peer spelling this replaced. That is what
     * lets the notification path become conversation-keyed without re-filing a
     * single existing 1:1 thread.
     *
     * Unlike [convTerm] this joins with [PEER_SEP], because a key is read back
     * apart again ([convPeers]) while a term is only ever compared whole.
     */
    fun convKey(peers: List<String>): String = peers.sorted().joinToString(PEER_SEP)

    /** The participants a conversation key names. Inverse of [convKey]. */
    fun convPeers(key: String): List<String> =
        key.split(PEER_SEP).filter { it.isNotEmpty() }

    /**
     * The derived index term for a participant set — the port of `dmConvTerm`.
     *
     * Joined with NOTHING rather than with a separator: a term crosses a NIP-50
     * search string, whose parse ends a token at whitespace. Pubkeys are
     * fixed-width 64-char hex, so concatenating them is unambiguous.
     */
    fun convTerm(peers: List<String>, namespace: String = CONV_TERM): String =
        "$namespace:" + peers.sorted().joinToString("")

    /** The namespace every rumor of a conversation is filed under. */
    const val CONV_TERM = "conv"

    /**
     * The namespace holding only the rumors worth listing (chat and file), so
     * `distinct:convmsg` can name the newest MESSAGE of every conversation
     * without any engine reading a rumor's kind. See `DM_MSG_TERM`.
     */
    const val MSG_TERM = "convmsg"

    /**
     * The same, restricted to messages the VIEWER sent — "conversations I have
     * written in", which is what tells the notification path that a sender is not
     * a stranger. See `DM_MINE_TERM`.
     */
    const val MINE_TERM = "convmine"

    /**
     * The [TermPolicy] for a `dm17:<self>` tenant: each rumor filed under the
     * one conversation it belongs to, in every namespace it qualifies for.
     *
     * `self` comes from the TENANT ID, which is the whole shape of the
     * contract — [SqliteArmadaDb] never interprets a tenant id or a term, and
     * this is the layer that spells `dm17:` in the first place.
     *
     * Exactly one term per namespace, which `distinct:` requires: a rumor that
     * was the newest of two groups could only ever be returned once.
     *
     * The service writes through this while the app is dead, so a message that
     * arrives then is already in the conversation list's index when the app
     * opens — which is the whole reason a policy binds to the tenant.
     */
    fun termsOf(rumor: Rumor, tenantId: String): List<String> {
        val self = tenantSelf(tenantId) ?: return emptyList()
        val peers = peersOf(rumor, self) ?: return emptyList()

        val terms = mutableListOf(convTerm(peers))
        if (rumor.kind in MESSAGE_KINDS) {
            terms.add(convTerm(peers, MSG_TERM))
            if (rumor.pubkey == self) terms.add(convTerm(peers, MINE_TERM))
        }
        return terms
    }

    /** The pubkey in a `dm17:<self>` tenant id, or null for any other id. */
    fun tenantSelf(tenantId: String): String? =
        if (tenantId.startsWith(TENANT_PREFIX)) tenantId.substring(TENANT_PREFIX.length) else null

    private const val TENANT_PREFIX = "dm17:"

    /**
     * Whether an opened DM rumor may be stored.
     *
     * What IS stored is the rumor, untouched — nothing is injected on the way
     * in. A rumor's tags are the bytes its id commits to, so bookkeeping
     * written into them makes the stored row something its sender never signed,
     * and makes the store's own idea of a conversation forgeable by anyone who
     * spells that tag themselves. The conversation is derived on read instead,
     * from the author and `p` tags ([peersOf]).
     *
     * Refused when:
     *
     *  - it isn't a DM-plane kind (a typing signal, or a foreign rumor kind
     *    that arrived in a DM wrap — Concord invites travel this way);
     *  - its NIP-40 deadline has passed. This is the single choke point every
     *    writer goes through, which is why the check lives here and not only at
     *    the point of decryption: a disappearing message that arrives late is
     *    simply never stored;
     *  - it has no attributable conversation, so no read could ever surface it
     *    in a thread.
     */
    fun storable(self: String, rumor: Rumor, now: Long): Boolean {
        if (self.isEmpty()) return false
        if (rumor.kind !in KINDS) return false
        if (isExpired(rumor.tags, now)) return false
        return peersOf(rumor, self) != null
    }
}

File: ios/ArmadaNotify/Sources/ArmadaNotify/Dm17.swift
import Foundation

/// A fully-opened, verified NIP-17 rumor, attributed to its conversation.
struct OpenedDm {
    let rumorId: String
    /// Verified author — the seal's signer, which equals the rumor's pubkey.
    let author: String
    let kind: Int
    let content: String
    let tags: [[String]]
    let createdAt: Int
    /// Everyone in the conversation except the viewer, sorted.
    let peers: [String]
    /// Backward-compatible shorthand for a 1:1's sole peer.
    var peer: String { peers[0] }
}

/// NIP-17 gift-wrap opening: `wrap(1059) → seal(13) → rumor`.
///
/// A port of `openDmWrap` in `src/lib/nip17/protocol.ts`, refusal for refusal.
/// That matters more than sharing any particular line of it: a notification
/// path that checked FEWER things than the app would be a second, laxer reader
/// of the same bytes, and the one that runs while nobody is watching.
enum Dm17 {

    static let kindChat = 14
    static let kindFile = 15
    static let kindSeal = 13
    static let kindWrap = 1059

    /// Rumors claiming to be further in the future than this are refused.
    private static let maxFutureSkewSecs = 3600

    /// Open a kind-1059 wrap addressed to `self`, or nil for anything the
    /// decryptor cannot open, that is malformed, or that has already expired.
    ///
    /// The checks, in order:
    ///   1. the wrap decrypts to us from `wrap.pubkey` → a kind-13 seal;
    ///   2. the seal decrypts to us from `seal.pubkey` → the rumor;
    ///   3. the rumor's claimed pubkey equals the seal's signer (NIP-59
    ///      anti-spoofing). No Schnorr verify is needed for that: NIP-44's AEAD
    ///      means a seal that decrypted was authenticated to US by its author;
    ///   4. the rumor's id, when it claims one, is its NIP-01 hash;
    ///   5. a NIP-40 deadline that has passed on the wrap, the seal OR the
    ///      rumor rejects the whole envelope. All three, because relays are not
    ///      trusted to have dropped it and a sender can strip the outer tag.
    ///
    /// Whether the sender turns out to be the viewer is NOT decided here — that
    /// is presentation policy, and the caller has to tell it apart from a wrap
    /// that simply would not open.
    static func open(
        wrap: NostrEvent,
        decryptor: Nip44Decryptor,
        self selfPubkey: String,
        now: Int = Int(Date().timeIntervalSince1970)
    ) -> OpenedDm? {
        guard wrap.kind == kindWrap else { return nil }
        guard !Nip40.isExpired(wrap.tags, now: now) else { return nil }

        guard let sealJson = decryptor.decrypt(pubkey: wrap.pubkey, ciphertext: wrap.content),
              let seal = NostrEvent.parse(json: sealJson),
              seal.kind == kindSeal,
              !Nip40.isExpired(seal.tags, now: now)
        else { return nil }

        guard let rumorJson = decryptor.decrypt(pubkey: seal.pubkey, ciphertext: seal.content),
              let rumor = NostrEvent.parse(json: rumorJson)
        else { return nil }

        guard rumor.pubkey == seal.pubkey else { return nil }
        guard rumor.createdAt <= now + maxFutureSkewSecs else { return nil }
        guard !Nip40.isExpired(rumor.tags, now: now) else { return nil }

        let computedId = rumor.computedId
        if let claimed = rumor.id, claimed != computedId { return nil }

        guard let peers = peersOf(rumor, self: selfPubkey) else { return nil }

        return OpenedDm(
            rumorId: computedId,
            author: rumor.pubkey,
            kind: rumor.kind,
            content: rumor.content,
            tags: rumor.tags,
            createdAt: rumor.createdAt,
            peers: peers
        )
    }

    /// Everyone involved except `self`, sorted. Matches `dmPeersOf` in the web
    /// client so both notification readers derive the same group room key.
    static func peersOf(_ rumor: NostrEvent, self selfPubkey: String) -> [String]? {
        var recipients = Set<String>()
        for tag in rumor.tags where tag.first == "p" && tag.count > 1 && !tag[1].isEmpty {
            recipients.insert(tag[1])
        }

        if rumor.pubkey != selfPubkey {
            recipients.insert(rumor.pubkey)
            recipients.remove(selfPubkey)
            return recipients.isEmpty ? nil : recipients.sorted()
        }

        guard !recipients.isEmpty else { return nil }
        recipients.remove(selfPubkey)
        return recipients.isEmpty ? [selfPubkey] : recipients.sorted()
    }

    /// Backward-compatible shorthand used by the existing 1:1 vectors.
    static func peerOf(_ rumor: NostrEvent, self selfPubkey: String) -> String? {
        peersOf(rumor, self: selfPubkey)?.first
    }
}
